### PR TITLE
Internal server error if there are multiple matching callbacks

### PR DESCRIPTION
**Pull request by frobijn**
## Description

If a developer has created multiple methods of controller(s) that match the same web request, an internal server error is generated with the names of the matching methods.

Plus: .editorconfig added. This also changes the header comment of the modified code files.

## Motivation and Context

It is a protection against coding errors. In the current code, if multiple callbacks match, both are executed with results in an exception in an unexpected place in the code. With the change applied, it is impossible to miss that there is a problem.

Assuming a developer/tester tests new or modified code the internal server error will not be generated in a production situation, so no data is leaked that may be considered a security risk.

I list it as "improvement", as one type of error (exception in the second callback) is replaced by another (internal server error).

## How Has This Been Tested?

Postman, as prescribed.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

## Summary by CodeRabbit

- **New Features**
	- Introduced multiple new test cases to enhance API testing, including `SimpleRouteController_MultipleCallback` and various scenarios under the `MixedController` category.
	- Added new variables to support enhanced test cases for authentication methods.

- **Bug Fixes**
	- Improved error handling to ensure appropriate server responses for specific scenarios, including validation of status codes and response content.

### CODE REVIEW

File: nanoFramework.WebServer/WebServer.cs [hunk 1]
```diff
@@ -537,35 +534,69 @@ private void StartListener()
...
+                            }
+                            else
+                            {
+                                multipleCallback ??= $"Multiple matching callbacks: {selectedRoute.Callback.DeclaringType.FullName}.{selectedRoute.Callback.Name}";
+                                multipleCallback += $", {route.Callback.DeclaringType.FullName}.{route.Callback.Name}";
+                            }
+                        }
 
-                            isRoute = true;
+                        if (multipleCallback is not null)
```
Comment (Ellerbach) [inline nanoFramework.WebServer/WebServer.cs:557]: Can this be done on DEBUG only? a bit like in the full .NET where errors appears only in debug mode?

Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:557]: Well, at the same time, it is compiled in Released mode in the nuget

Comment (frobijn) [reply nanoFramework.WebServer/WebServer.cs:557]: Yes  that's the reason. The message for the internal server error is the only communication channel available. AFAIK there is no way to control the output of the class library depending on the configuration of the main program in .NET, so I guess that's the same for nanoFramework. In .NET there's trace as alternative for debug messages, but not in nanoframework.

File: nanoFramework.WebServer/WebServer.cs [hunk 2]
```diff
@@ -537,35 +534,69 @@ private void StartListener()
...
+                        {
+                            multipleCallback += ".";
+                            context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
+                            OutPutStream(context.Response, multipleCallback);
 
+                            HandleContextResponse(context);
+                        }
+                        else if (selectedRoute is null)
+                        {
+                            if (CommandReceived != null)
```
Comment (Ellerbach) [inline nanoFramework.WebServer/WebServer.cs:567]: You still need to check auth. So, you should have to adjust with the previous logic.

File: nanoFramework.WebServer/WebServer.cs (hunk 1 above)
Comment (Ellerbach) [reply nanoFramework.WebServer/WebServer.cs:557]: ok, fine with that then. I don't want to add another element in the lib. Let's then display this error message when multiple functions with the same route are found.

**Review by josesimoes — COMMENTED** [review]
@frobijn let's have atomic PRs please. Change only what is meant to be changed on the PR. In this case the fix for the multiple matching callbacks.

If there is an issue with the way wifi credentials are being stored, lets discuss a way to address that and add those changes on another PR.

**Comment by frobijn** [comment]
> @frobijn let's have atomic PRs please. Change only what is meant to be changed on the PR. In this case the fix for the multiple matching callbacks.
> 
> If there is an issue with the way wifi credentials are being stored, lets discuss a way to address that and add those changes on another PR.

But that should then have been in a PR before this one. I made the change of multiple callbacks before having issues with the WiFi credentials. If I would have to make the WiFi change before I could continue with the multiple callback change. Given the wait times between initiating a PR and having it closed, that makes it quite hard to do some small work. I would almost have to work on as many branches as there are code changes. Please please please allow for small improvements that make life easier for a contributor.

**Comment by josesimoes** [comment]
In this case it's a simple matter of not including the changes with wifi.cs, nothing complicated.

Please understand that this is not to make developers life complicated, is to improve project maintainability. 
If we need to revert this set of changes for any reason, and if includes _other_ unrelated changes, it's more work and one has to remember to address that.
Plus, change log is generated from the PR title. If that includes multiple unrelated changes it has impact on formatting and clarity.

FYI: I've just extracted these changes and added them in #270 . Updating this branch now so these are not part of the PR anymore. 

Our contribution guideline suggests that for breaking or large changes, these are discussed beforehand. That is to prevent waste of time/work and even frustration. No one likes to see it's work changed or the effort dismissed. No one here wants to bash developers willing to contribute. Quite the oposite. Still there are aspects that require a broader and 10.000ft view that (most likely) only maintainers have.

Also know that for simple changes and fixes, you can do that right on the github webpage. No need to branch, github offers to start a PR for you.

**Comment by josesimoes** [comment]
I understand your point. Sometimes it can hinder productivity, but it's required for the greater good. 😉 
Merging and rebasing when one is working on the same change set is usually not painful. Things do get complicated when there is active development and even more if several developers are working in the same files. That requires a more frequent coordination. Thank you for your understanding! 👍🏻

File: README.md [hunk 3]
```diff
@@ -196,6 +196,50 @@ With the previous example the following happens:
...
+    }
+
+    [Route("sameroute")]
+    public void None(WebServerEventArgs e)
+    {
+        WebServer.OutPutStream(e.Context.Response, "sameroute: Public");
+    }
+}
+```
+The webserver selects the route for a request:
```
Comment (Ellerbach) [inline README.md:233]:
```suggestion

The webserver selects the route for a request:
```

File: README.md [hunk 4]
```diff
@@ -196,6 +196,50 @@ With the previous example the following happens:
...
+
+- If there are no matching methods, a not-found response (404) is returned.
+- If authentication information is passed in the header of the request, then only methods that require authentication are considered. If one of the method's credentials matches the credentials passed in the request, that method is called. Otherwise a non-authorized response (401) will be returned.
+- If no authentication information is passed in the header of the request:
+	- If one of the methods does not require authentication, that method is called.
+	- Otherwise a non-authorized response (401) will be returned. If one of the methods requires basic authentication, the `WWW-Authenticate` header is included to request credentials.
+
+If two or more methods match the authentication method and credentials of the request, an internal server error is returned with a list of the methods.
```
Comment (Ellerbach) [inline README.md:241]: then from the previous posted experience, can you be a bit more explicit on what will happen?

**Review by Ellerbach — COMMENTED** [review]
Looks great now. Few comments on the readme, then I think we will be good to go

**Comment by frobijn** [comment]
I tried to clean up my local git repository and create a cleaner MultipleCallbacks branch. I'm a bit surprised that this PR got closed by it. A new PR is on its way.

File: README.md (hunk 4 above)
Comment (frobijn) [reply README.md:241]: In the new PR, the line last line is replaced by:

The webserver does not support more than one matching method. Calling multiple methods most likely results in an exception as a subsequent method tries to modify a response that is already processed by the first method. The webserver does not know what to do and returns an internal server error (500). The body of the response lists the matching methods.

Having multiple matching methods is considered a programming error. One way this occurs is if two methods in a controller accidentally have the same route. Returning an internal server error with the names of the methods makes it easy to discover the error. It is expected that the error is discovered and fixed in testing. Then the internal error will not occur in the application that is deployed to a device.

**Comment by josesimoes** [comment]
@frobijn this PR was closed the moment you've deleted the branch that originated it. It can be restored and reopened. If you prefer.

**Pull request closed without merging**